### DISCUSSION TITLE
-colors 255 always applying 256

### DISCUSSION THREAD

**Discussion by Vanir-Labs** [ImageMagick/ImageMagick#7030] [Help]
### ImageMagick version

7.1.1-26

### Operating system, version and so on

Windows 10

### Description

Hello!
I'm trying to use the -colors 255 but it keep making it 256 colours. My program needs to have exactly 255 colours with 8 bit depth.
Any ideas how I can fix this?

`magick "C:\textures\1bmp" -format bmp -alpha off -type palette -colors 255 -depth 8 -compress none BMP3:"C:\textures2\1.bmp"`

**Comment by snibgo** [comment]
With a sample image, I get 255 colors, tested with `-unique-colors`.

Please show your input image that gives you 256 colours. Please also show how you know it has 256 colours.

**Reply by Vanir-Labs** [reply]
Thanks for testing this. I'll give this a try, for now to answer your question I use a metadata extraction web app. 
https://www.metadata2go.com/view-metadata
Image stats are:
`num_colors
255
num_important_colors
255`
[texture.zip](https://github.com/ImageMagick/ImageMagick/files/13934641/texture.zip)

**Reply by Vanir-Labs** [reply]
When I do `-unique-colors` it does this to my output image.
![image](https://github.com/ImageMagick/ImageMagick/assets/82656075/aea5227e-4420-44b9-ac74-8da7d7ae0579)

**Comment by snibgo** [comment]
> num_colors 255 num_important_colors 255

So, you also get only 255 colors?

Your zipped file nam_2ob.bmp contains only 255 colours, as we can show like this:
```
C:\web\im>%IMG7%magick nam_2ob.bmp -unique-colors info:
nam_2ob.bmp BMP3 255x1 255x1+0+0 8-bit sRGB 66612B 0.000u 0:00.000
```
This makes the image Nx1 pixels, where each input colour occurs only once. So the width of the result is the number of unique colours of the input image.

The `-colors 255` operation can't add new colours.

**Reply by Vanir-Labs** [reply]
I understand it says it can add colours to it, but it's clearly adding 1 colour somewhere somwhow. Here's the image after I execute the initial colours 255 command in the rest of that line.
[texture256.zip](https://github.com/ImageMagick/ImageMagick/files/13934755/texture256.zip)

**Comment by snibgo** [comment]
Your "texture256.zip" file has 209 unique colours:
```
C:\web\im>%IMG7%magick texture256\nam_2ob.bmp -unique-colors info:
texture256\nam_2ob.bmp BMP 209x1 209x1+0+0 8-bit sRGB 263306B 0.000u 0:00.000
```

**Reply by Vanir-Labs** [reply]
The metadata and my game say otherwise. My game can't read 256 colours. only 255 and below. this image errors every time it loads.
![image](https://github.com/ImageMagick/ImageMagick/assets/82656075/eac86e22-e923-42c1-a7e1-2207eb44ba59)

How can I force colours to be 255 regardless of what the image has?

**Reply by Vanir-Labs** [reply]
I think it has to do with it not converting to BMPv3 and keeping the BMPv5. I'll continue to adjust and run tests.

**Comment by snibgo** [comment]
Those are fields in the BMP header. `num_colors` is the number of entries in the palette, not the number of colors in the image. `num_important` means nothing. See https://en.wikipedia.org/wiki/BMP_file_format
